src/checkpoint/traits/serializable_traits.h

```diff
@@ -117,6 +118,10 @@ struct SerializableTraits {
   using constructor_t = decltype(U());
   using has_default_constructor = detection::is_detected<constructor_t, T>;
 
+  template <typename U>
+  using tagged_constructor_t = decltype(U(std::declval<SERIALIZE_CONSTRUCT_TAG>()));
+  using has_tagged_constructor = detection::is_detected<tagged_constructor_t, T>;
+
   template <typename U>
   using reconstruct_t = decltype(U::reconstruct(std::declval<void*>()));
   using has_reconstruct =
```

Comment:
I think this needs to be stricter, to avoid the case where a parent class (direct or not) has a `reconstruct` member, but `U` does not.

One consideration is that we need `is_same<T, reconstruct_t>` rather than convertibility - otherwise, we're calling the wrong serializer.

Another consideration is something along the lines of 
```
no_base_type or not has_reconstruct<BaseType> or &static_cast<U::*(void*)>(U::reconstruct) != &static_cast<U::*(void*)>(Base::reconstruct)
```
I.e. that if this is a derived class, and some base defines a `reconstruct` method, that's not the one we're going to try to call.

Reply:
Turns out that the return type of reconstruct makes it distinct. Even with `is_detected_convertible` this checks for an **implicit** conversion from Base to Derived. The only way that could potentially be valid is with a constructor in Derived that takes Base as an argument. However, since this is checking convertibility on a reference to derived, it even that doesn't trigger it.

Reply:
I even tried getting it to fail with `operator BaseT&() & { return *this; }`, but this didn't trigger it either. Note, with the constructor implicit converter I got `std::is_convertible<BaseT, DerviedT>::value` to be true, but not `std::is_convertible<BaseT&, DerviedT&>::value`

Reply:
I've changed this to `is_detected_exact` to be super safe and added a test to ensure it works in this case.